File: src/test/java/com/jore/epoc/SimulationBuilder.java
package com.jore.epoc;

import java.time.YearMonth;
import java.util.ArrayList;
import java.util.List;

import com.jore.datatypes.currency.Currency;
import com.jore.datatypes.money.Money;
import com.jore.datatypes.percent.Percent;
import com.jore.epoc.bo.Company;
import com.jore.epoc.bo.Market;
import com.jore.epoc.bo.MarketSimulation;
import com.jore.epoc.bo.Simulation;
import com.jore.epoc.bo.accounting.FinancialAccounting;
import com.jore.epoc.bo.orders.AbstractSimulationOrder;
import com.jore.epoc.bo.orders.AdjustCreditLineOrder;
import com.jore.epoc.bo.orders.BuildFactoryOrder;
import com.jore.epoc.bo.orders.BuildStorageOrder;
import com.jore.epoc.bo.orders.BuyRawMaterialOrder;
import com.jore.epoc.bo.orders.ChangeAmountAndPriceOrder;
import com.jore.epoc.bo.orders.CreditEventDirection;
import com.jore.epoc.bo.orders.EnterMarketOrder;
import com.jore.epoc.bo.orders.IncreaseProductivityOrder;
import com.jore.epoc.bo.orders.IncreaseQualityOrder;
import com.jore.epoc.bo.orders.MarketingCampaignOrder;
import com.jore.epoc.bo.settings.EpocSettings;

public class SimulationBuilder {
    private static final String CHF = "CHF";
    private static int ID = 0;

    public static SimulationBuilder builder() {
        return new SimulationBuilder();
    }

    // System parameters
    private Percent interestRate = Percent.of("5%");
    private Currency baseCurrency = Currency.getInstance(CHF);
    private Money constructionCostsPerUnit = Money.of(CHF, 100);
    private Money constructionCosts = Money.of(CHF, 1000000);
    private Money rawMaterialUnitPrice = Money.of(CHF, 300);
    private Money factoryConstructionCosts = Money.of(CHF, 1000000);
    private Percent demandCurveLowerPercent = Percent.of("80%");
    private Money demandCurveLowerPrice = Money.of(CHF, 500);
    private Percent demandCurveHigherPercent = Percent.of("20%");
    private Money demandCurveHigherPrice = Money.of(CHF, 1200);
    private int productLifecycleDuration = 100;
    private Integer timeToBuildStorage = 0;
    private Integer timeToBuildFactory = 0;
    private Integer monthlyCapacityPerProductionLine = 100;
    private Money factoryLaborCost = Money.of(CHF, 500000);
    private String marketName = "Switzerland";
    private Money marketDistributionCost = Money.of(CHF, 2000000);
    private int marketSize = 1000000;
    private Money buildingMaintenanceCost = Money.of(CHF, 10000);
    private Percent depreciationRate = Percent.of("15%");
    private Money headquarterCost = Money.of(CHF, 1500000);
    private Money inventoryManagementCost = Money.of(CHF, 500000);
    private Money marketEntryCost = Money.of(CHF, 400000);
    private Money productionCost = Money.of(CHF, 30);
    private Integer simulationPassiveSteps = 2;
    private Money pricePerPointQuality = Money.of("CHF", 200000);
    private Money pricePerMarketingCampaign = Money.of("CHF", 500000);
    private Money pricePerProductivityPoint = Money.of("CHF", 500000);
    private Percent factorDiscountRate = Percent.of("10%");
    // To be set in application
    private Money initialOfferedPrice = Money.of(CHF, 800);
    private int initialIntendedSale = 1000;
    private Integer productionLines = 10;
    private Money productionLineConstructionCosts = Money.of(CHF, 100000);
    private YearMonth simulationStart = YearMonth.of(2020, 1);
    private String name = "Unnamed Company";
    private String simulationName = "Unnamed Simulation";
    private Integer numberOfSimulationSteps = 24;
    private List<AbstractSimulationOrder> orders = new ArrayList<>();
    private Simulation simulation;

    public Company build() {
        Company result = new Company();
        result.setId(ID++);
        result.setName(name);
        if (simulation == null) {
            createSimulation();
            Market market = createMarket();
            MarketSimulation marketSimulation = createMarketSimulation();
            marketSimulation.setMarket(market);
            simulation.addMarketSimulation(marketSimulation);
        }
        simulation.addCompany(result);
        FinancialAccounting accounting = new FinancialAccounting();
        accounting.setId(ID++);
        accounting.setBaseCurrency(baseCurrency);
        result.setFinancialAccounting(accounting);
        for (AbstractSimulationOrder order : orders) {
            result.addSimulationOrder(order);
            if (order instanceof EnterMarketOrder enterMarketOrder) {
                enterMarketOrder.setMarketSimulation(simulation.getMarketSimulations().get(0));
            }
            if (order instanceof ChangeAmountAndPriceOrder changeAmountAndPriceOrder) {
                changeAmountAndPriceOrder.setMarket(simulation.getMarketSimulations().get(0).getMarket());
            }
        }
        return result;
    }

    public SimulationBuilder buildFactory(YearMonth executionMonth) {
        BuildFactoryOrder order = new BuildFactoryOrder();
        order.setExecutionMonth(executionMonth);
        order.setConstructionCost(factoryConstructionCosts);
        order.setConstructionCostPerLine(productionLineConstructionCosts);
        order.setProductionLines(productionLines);
        order.setTimeToBuild(timeToBuildFactory);
        order.setMonthlyCapacityPerProductionLine(monthlyCapacityPerProductionLine);
        order.setProductionLineLaborCost(factoryLaborCost);
        orders.add(order);
        return this;
    }

    public SimulationBuilder buildStorage(YearMonth executionMonth, int storageCapacity) {
        BuildStorageOrder order = new BuildStorageOrder();
        order.setExecutionMonth(executionMonth);
        order.setCapacity(storageCapacity);
        order.setConstructionCostPerUnit(constructionCostsPerUnit);
        order.setConstructionCost(constructionCosts);
        order.setTimeToBuild(timeToBuildStorage);
        order.setInventoryManagementCost(inventoryManagementCost);
        orders.add(order);
        return this;
    }

    public SimulationBuilder buyRawMaterial(YearMonth executionMonth, int amount) {
        BuyRawMaterialOrder order = new BuyRawMaterialOrder();
        order.setExecutionMonth(executionMonth);
        order.setAmount(amount);
        order.setUnitPrice(rawMaterialUnitPrice);
        orders.add(order);
        return this;
    }

    public SimulationBuilder changeAmountAndPriceOrder(YearMonth executionMonth, int amount, Money price) {
        ChangeAmountAndPriceOrder order = new ChangeAmountAndPriceOrder();
        order.setExecutionMonth(executionMonth);
        order.setIntentedSales(amount);
        order.setOfferedPrice(price);
        orders.add(order);
        return this;
    }

    public SimulationBuilder enterMarket(YearMonth executionMonth) {
        EnterMarketOrder order = new EnterMarketOrder();
        order.setExecutionMonth(executionMonth);
        order.setOfferedPrice(initialOfferedPrice);
        order.setIntentedProductSale(initialIntendedSale);
        order.setEnterMarktCost(marketEntryCost);
        orders.add(order);
        return this;
    }

    public SimulationBuilder increaseCreditLine(YearMonth executionMonth, Money increaseAmount) {
        AdjustCreditLineOrder order = new AdjustCreditLineOrder();
        order.setExecutionMonth(executionMonth);
        order.setAmount(increaseAmount);
        order.setDirection(CreditEventDirection.INCREASE);
        order.setInterestRate(interestRate);
        orders.add(order);
        return this;
    }

    public SimulationBuilder increaseProductivity(YearMonth executionMonth, Money increaseProductivityAmount) {
        IncreaseProductivityOrder order = new IncreaseProductivityOrder();
        order.setExecutionMonth(executionMonth);
        order.setAmount(increaseProductivityAmount);
        orders.add(order);
        return this;
    }

    public SimulationBuilder increaseQuality(YearMonth executionMonth, Money increaseQualityAmount) {
        IncreaseQualityOrder order = new IncreaseQualityOrder();
        order.setExecutionMonth(executionMonth);
        order.setAmount(increaseQualityAmount);
        orders.add(order);
        return this;
    }

    public SimulationBuilder laborForce(int marketSize) {
        this.marketSize = marketSize;
        return this;
    }

    public SimulationBuilder marketingCampaign(YearMonth executionMonth, Money marketingCampaignAmount) {
        MarketingCampaignOrder order = new MarketingCampaignOrder();
        order.setExecutionMonth(executionMonth);
        order.setAmount(marketingCampaignAmount);
        orders.add(order);
        return this;
    }

    public SimulationBuilder marketSize(int marketSize) {
        this.marketSize = marketSize;
        return this;
    }

    public SimulationBuilder name(String name) {
        this.name = name;
        return this;
    }

    public SimulationBuilder numberOfSimulationSteps(Integer numberOfSimulationSteps) {
        this.numberOfSimulationSteps = numberOfSimulationSteps;
        return this;
    }

    public SimulationBuilder passiveSteps(int passiveSteps) {
        this.simulationPassiveSteps = passiveSteps;
        return this;
    }

    public SimulationBuilder simulation(Simulation simulation) {
        this.simulation = simulation;
        return this;
    }

    public SimulationBuilder simulationName(String simulationName) {
        this.simulationName = simulationName;
        return this;
    }

    private Market createMarket() {
        Market market = new Market();
        market.setId(ID++);
        market.setName(marketName);
        market.setMarketSize(marketSize);
        market.setDistributionCost(marketDistributionCost);
        market.setCostToEnterMarket(marketEntryCost);
        return market;
    }

    private MarketSimulation createMarketSimulation() {
        MarketSimulation marketSimulation = new MarketSimulation();
        marketSimulation.setId(ID++);
        marketSimulation.setStartMonth(simulationStart); // TODO has this really to be set manually?
        marketSimulation.setLowerPercent(demandCurveLowerPercent);
        marketSimulation.setLowerPrice(demandCurveLowerPrice);
        marketSimulation.setHigherPercent(demandCurveHigherPercent);
        marketSimulation.setHigherPrice(demandCurveHigherPrice);
        marketSimulation.setProductLifecycleDuration(productLifecycleDuration);
        return marketSimulation;
    }

    private void createSimulation() {
        EpocSettings settings = new EpocSettings();
        simulation = new Simulation();
        settings.addSetting(SettingBuilder.builder().settingKey(EpocSettings.PASSIVE_STEPS).valueText(simulationPassiveSteps.toString()).build());
        settings.addSetting(SettingBuilder.builder().settingKey(EpocSettings.PRICE_PER_POINT_QUALITY).valueText(pricePerPointQuality.toString()).build());
        settings.addSetting(SettingBuilder.builder().settingKey(EpocSettings.PRICE_PER_MARKETING_CAMPAIGN).valueText(pricePerMarketingCampaign.toString()).build());
        settings.addSetting(SettingBuilder.builder().settingKey(EpocSettings.PRICE_PER_PRODUCTIVITY_POINT).valueText(pricePerProductivityPoint.toString()).build());
        settings.addSetting(SettingBuilder.builder().settingKey(EpocSettings.FACTOR_DISCOUNT_RATE).valueText(factorDiscountRate.toString()).build());
        simulation.setSettings(settings);
        simulation.setId(ID++);
        simulation.setName(simulationName);
        simulation.setStartMonth(simulationStart);
        simulation.setNrOfMonths(numberOfSimulationSteps);
        simulation.setInterestRate(interestRate);
        simulation.setBuildingMaintenanceCost(buildingMaintenanceCost);
        simulation.setDepreciationRate(depreciationRate);
        simulation.setHeadquarterCost(headquarterCost);
        simulation.setProductionCost(productionCost);
    }
}

File: EpocServer/Tests/AppTests/SimulationTests/SimulationBuilder.swift
import Foundation

/// Builder to construct complete simulation setups for testing.
/// Equivalent to `com.jore.epoc.SimulationBuilder`.
public class SimulationBuilder {
    private static let CHF = "CHF"
    private static var nextId: Int64 = 0

    private static func nextID() -> Int64 {
        let id = nextId
        nextId += 1
        return id
    }

    public static func builder() -> SimulationBuilder {
        SimulationBuilder()
    }

    // System parameters
    private var interestRate = Percent.of("5%")
    private var baseCurrency = Currency.getInstance("CHF")
    private var constructionCostsPerUnit = Money.of(CHF, 100)
    private var constructionCosts = Money.of(CHF, 1000000)
    private var rawMaterialUnitPrice = Money.of(CHF, 300)
    private var factoryConstructionCosts = Money.of(CHF, 1000000)
    private var demandCurveLowerPercent = Percent.of("80%")
    private var demandCurveLowerPrice = Money.of(CHF, 500)
    private var demandCurveHigherPercent = Percent.of("20%")
    private var demandCurveHigherPrice = Money.of(CHF, 1200)
    private var productLifecycleDuration = 100
    private var timeToBuildStorage = 0
    private var timeToBuildFactory = 0
    private var dailyCapacityPerProductionLine = 4
    private var factoryLaborCost = Money.of(CHF, 500000)
    private var marketNameValue = "Switzerland"
    private var marketDistributionCost = Money.of(CHF, 2000000)
    private var laborForceValue = 1000000
    private var buildingMaintenanceCostValue = Money.of(CHF, 10000)
    private var depreciationRateValue = Percent.of("15%")
    private var headquarterCostValue = Money.of(CHF, 1500000)
    private var inventoryManagementCostValue = Money.of(CHF, 500000)
    private var marketEntryCost = Money.of(CHF, 400000)
    private var productionCostValue = Money.of(CHF, 30)
    private var simulationPassiveSteps = 3
    private var pricePerPointQuality = Money.of(CHF, 200000)
    private var pricePerMarketingCampaign = Money.of(CHF, 500000)
    private var pricePerProductivityPoint = Money.of(CHF, 500000)
    private var factorDiscountRate = Percent.of("10%")
    // To be set in application
    private var initialOfferedPrice = Money.of(CHF, 800)
    private var initialIntendedSale = 1000
    private var productionLinesValue = 10
    private var productionLineConstructionCosts = Money.of(CHF, 100000)
    private var simulationStart = YearMonth.of(2020, 1)
    private var nameValue = "Unnamed Company"
    private var simulationNameValue = "Unnamed Simulation"
    private var numberOfSimulationStepsValue = 24
    private var orders: [any SimulationOrder] = []
    private var simulationValue: Simulation?

    @discardableResult
    public func name(_ name: String) -> SimulationBuilder {
        self.nameValue = name
        return self
    }

    @discardableResult
    public func simulationName(_ simulationName: String) -> SimulationBuilder {
        self.simulationNameValue = simulationName
        return self
    }

    @discardableResult
    public func numberOfSimulationSteps(_ numberOfSimulationSteps: Int) -> SimulationBuilder {
        self.numberOfSimulationStepsValue = numberOfSimulationSteps
        return self
    }

    @discardableResult
    public func passiveSteps(_ passiveSteps: Int) -> SimulationBuilder {
        self.simulationPassiveSteps = passiveSteps
        return self
    }

    @discardableResult
    public func laborForce(_ laborForce: Int) -> SimulationBuilder {
        self.laborForceValue = laborForce
        return self
    }

    @discardableResult
    public func simulation(_ simulation: Simulation) -> SimulationBuilder {
        self.simulationValue = simulation
        return self
    }

    // MARK: - Order Builders

    @discardableResult
    public func increaseCreditLine(_ executionMonth: YearMonth, _ increaseAmount: Money) -> SimulationBuilder {
        let order = AdjustCreditLineOrder()
        order.executionMonth = executionMonth
        order.amount = increaseAmount
        order.direction = .increase
        order.interestRate = interestRate
        orders.append(order)
        return self
    }

    @discardableResult
    public func buildFactory(_ executionMonth: YearMonth) -> SimulationBuilder {
        let order = BuildFactoryOrder()
        order.executionMonth = executionMonth
        order.constructionCost = factoryConstructionCosts
        order.constructionCostPerLine = productionLineConstructionCosts
        order.productionLines = productionLinesValue
        order.timeToBuild = timeToBuildFactory
        order.dailyCapacityPerProductionLine = dailyCapacityPerProductionLine
        order.productionLineLaborCost = factoryLaborCost
        orders.append(order)
        return self
    }

    @discardableResult
    public func buildStorage(_ executionMonth: YearMonth, _ storageCapacity: Int) -> SimulationBuilder {
        let order = BuildStorageOrder()
        order.executionMonth = executionMonth
        order.capacity = storageCapacity
        order.constructionCostPerUnit = constructionCostsPerUnit
        order.constructionCost = constructionCosts
        order.timeToBuild = timeToBuildStorage
        order.inventoryManagementCost = inventoryManagementCostValue
        orders.append(order)
        return self
    }

    @discardableResult
    public func buyRawMaterial(_ executionMonth: YearMonth, _ amount: Int) -> SimulationBuilder {
        let order = BuyRawMaterialOrder()
        order.executionMonth = executionMonth
        order.amount = amount
        order.unitPrice = rawMaterialUnitPrice
        orders.append(order)
        return self
    }

    @discardableResult
    public func enterMarket(_ executionMonth: YearMonth) -> SimulationBuilder {
        let order = EnterMarketOrder()
        order.executionMonth = executionMonth
        order.offeredPrice = initialOfferedPrice
        order.intentedProductSale = initialIntendedSale
        order.marketEntryCost = marketEntryCost
        orders.append(order)
        return self
    }

    @discardableResult
    public func changeAmountAndPriceOrder(_ executionMonth: YearMonth, _ amount: Int, _ price: Money) -> SimulationBuilder {
        let order = ChangeAmountAndPriceOrder()
        order.executionMonth = executionMonth
        order.intentedSales = amount
        order.offeredPrice = price
        orders.append(order)
        return self
    }

    @discardableResult
    public func increaseQuality(_ executionMonth: YearMonth, _ increaseQualityAmount: Money) -> SimulationBuilder {
        let order = IncreaseQualityOrder()
        order.executionMonth = executionMonth
        order.increaseQualityAmount = increaseQualityAmount
        orders.append(order)
        return self
    }

    @discardableResult
    public func increaseProductivity(_ executionMonth: YearMonth, _ increaseProductivityAmount: Money) -> SimulationBuilder {
        let order = IncreaseProductivityOrder()
        order.executionMonth = executionMonth
        order.increaseProductivityAmount = increaseProductivityAmount
        orders.append(order)
        return self
    }

    @discardableResult
    public func marketingCampaign(_ executionMonth: YearMonth, _ marketingCampaignAmount: Money) -> SimulationBuilder {
        let order = MarketingCampaignOrder()
        order.executionMonth = executionMonth
        order.marketingCampaignAmount = marketingCampaignAmount
        orders.append(order)
        return self
    }

    // MARK: - Build

    public func build() -> Company {
        let result = Company()
        result.id = SimulationBuilder.nextID()
        result.name = nameValue
        if simulationValue == nil {
            createSimulation()
            let market = createMarket()
            let marketSimulation = createMarketSimulation()
            marketSimulation.market = market
            simulationValue!.addMarketSimulation(marketSimulation)
        }
        simulationValue!.addCompany(result)
        let accounting = FinancialAccounting()
        accounting.id = SimulationBuilder.nextID()
        accounting.baseCurrency = baseCurrency
        result.accounting = accounting
        for order in orders {
            result.addSimulationOrder(order)
            if let enterMarketOrder = order as? EnterMarketOrder {
                enterMarketOrder.marketSimulation = simulationValue!.marketSimulations[0]
            }
            if let changeAmountAndPriceOrder = order as? ChangeAmountAndPriceOrder {
                changeAmountAndPriceOrder.market = simulationValue!.marketSimulations[0].market
            }
        }
        return result
    }

    // MARK: - Private

    private func createSimulation() {
        let settings = EpocSettings()
        simulationValue = Simulation()
        func addSetting(key: String, value: String) {
            let setting = EpocSetting()
            setting.settingKey = key
            setting.valueText = value
            settings.addSetting(setting)
        }
        addSetting(key: EpocSettings.PASSIVE_STEPS, value: "\(simulationPassiveSteps)")
        addSetting(key: EpocSettings.PRICE_PER_POINT_QUALITY, value: "\(pricePerPointQuality)")
        addSetting(key: EpocSettings.PRICE_PER_MARKETING_CAMPAIGN, value: "\(pricePerMarketingCampaign)")
        addSetting(key: EpocSettings.PRICE_PER_PRODUCTIVITY_POINT, value: "\(pricePerProductivityPoint)")
        addSetting(key: EpocSettings.FACTOR_DISCOUNT_RATE, value: "\(factorDiscountRate)")
        simulationValue!.settings = settings
        simulationValue!.id = SimulationBuilder.nextID()
        simulationValue!.name = simulationNameValue
        simulationValue!.startMonth = simulationStart
        simulationValue!.nrOfMonths = numberOfSimulationStepsValue
        simulationValue!.interestRate = interestRate
        simulationValue!.buildingMaintenanceCost = buildingMaintenanceCostValue
        simulationValue!.depreciationRate = depreciationRateValue
        simulationValue!.headquarterCost = headquarterCostValue
        simulationValue!.productionCost = productionCostValue
    }

    private func createMarket() -> Market {
        let market = Market()
        market.id = SimulationBuilder.nextID()
        market.name = marketNameValue
        market.laborForce = laborForceValue
        market.distributionCost = marketDistributionCost
        market.costToEnterMarket = marketEntryCost
        return market
    }

    private func createMarketSimulation() -> MarketSimulation {
        let marketSimulation = MarketSimulation()
        marketSimulation.id = SimulationBuilder.nextID()
        marketSimulation.startMonth = simulationStart
        marketSimulation.lowerPercent = demandCurveLowerPercent
        marketSimulation.lowerPrice = demandCurveLowerPrice
        marketSimulation.higherPercent = demandCurveHigherPercent
        marketSimulation.higherPrice = demandCurveHigherPrice
        marketSimulation.productLifecycleDuration = productLifecycleDuration
        return marketSimulation
    }
}
